WordPress/gutenberg
# Add PluginHeaderToolbar SlotFill to edit-post package.

## What?
This PR is an updated version of #17198 and is meant to address #16988.
I moved it to the main repo and updated to latest trunk as my fork was very outdated and this seemed to be the best way to move this forward.

## Why?
There are many plugins that have been inserting elements into the header via workaround and hacks. This would allow us to provide an elegant solution and to an extent place some guardrails on the UX

## How?
This PR introduces a new SlotFill location to the end of the toolbar that is only displayed when the view port is large (similar to the tools selector ). 

## Testing Instructions
1. Checkout this branch and build
2. introduce a plugin with the sample code below
3. Confirm that a new icon appears and is clickable as seen in the screeonshow


Sample Code:
```js
import { registerPlugin } from '@wordpress/plugins';
import { PluginHeaderToolbar } from '@wordpress/edit-post';

const MyHeaderToolbarPlugin = () => (
	<PluginHeaderToolbar
		className="plugin-header-toolbar-button"
		classContentName="plugin-header-toolbar-content"
		renderContent={ () => <div>Rendered content</div> }
	/>
);

registerPlugin( 'my-header-toolbar-plugin', {
	render: MyHeaderToolbarPlugin,
	icon: 'smiley',
} );
```

## Screenshots or screencast
![example](https://github.com/WordPress/gutenberg/assets/1259027/2db06a40-a028-48cd-85ba-187973f7bcda)
